common/tablefunctions.lua

```diff
@@ -104,11 +103,166 @@ if not table.invert then
+		for key, v in ipairs(tbl) do
+			if v == value then
+				found = key
+				break
+			end
+		end
+		if found then
+			table.remove(tbl, found)
+			return true
+		end
+
+		-- otherwise, try to handle the table normally and simply `nil` the value
+		found = table.findKey(tbl, value)
+		if found then
+			tbl[found] = nil
+			return true
+		end
+
+		return false
+	end
+
+	local function testTableRemoveValue()
+		local tests = {
+			sequence = { "a", "b", "c" }, -- indexes should be kept without any gaps for this one
+			notASequence = { "a", "b", [4] = "c" },
+			regularTable = { a = "a", b = "b", c = "c" },
+		}
+
+		for name, test in pairs(tests) do
+			Spring.Echo(string.format("[testTableRemoveValue] %s: %s", name, table.toString(test)))
+			for _, value in pairs({ "b", "c", "a" }) do
+				Spring.Echo(string.format("[testTableRemoveValue] remove: %s", value))
+				table.removeValue(test, value)
+				Spring.Echo(string.format("[testTableRemoveValue] %s: %s", name, table.toString(test)))
+			end
+		end
+	end
+
+	-- testTableRemoveValue()
+end
+
+if not table.shuffle then
+	---Shuffle sequence using Knuth (Fisher–Yates) algorithm.
+	---@param sequence any[] must be a Lua sequence (i.e. indexes form a contiguous sequence starting from 1), with the exception that we optionally allow starting from 0
+	---@param firstIndex? 0|1 first index in the sequence (optional, default: 1)
+	function table.shuffle(sequence, firstIndex)
+		firstIndex = firstIndex or 1
+		for i = firstIndex, #sequence - 2 + firstIndex do
+			local j = math.random(i, #sequence)
+			sequence[i], sequence[j] = sequence[j], sequence[i]
+		end
+	end
+
+	local function testTableShuffle()
+		local t = { "a", "b", "c" }
+		local t0 = { [0] = "a", "b", "c" }
+		local results = { abc = 0, acb = 0, bac = 0, bca = 0, cab = 0, cba = 0 }
+		local results0 = { abc = 0, acb = 0, bac = 0, bca = 0, cab = 0, cba = 0 }
+
+		for _ = 1, 10000000 do
+			table.shuffle(t)
+			table.shuffle(t0, 0)
+			local r = t[1] .. t[2] .. t[3]
+			local r0 = t0[0] .. t0[1] .. t0[2]
+			results[r] = results[r] + 1
+			results0[r0] = results0[r0] + 1
+		end
+
+		for k in pairs(results) do
+			Spring.Echo(string.format("[testTableShuffle] %s: [r] %d // [r0] %d", k, results[k], results0[k]))
+		end
+	end
+
+	-- testTableShuffle()
+end
+
+if not table.pairsByKeys then
+	---Alternative pairs-like function that traverse the table in the order of its
+	---keys. Straight from the docs https://www.lua.org/pil/19.3.html.
+	---@param t any
+	---@param f? any function passed to table.sort for sorting
+	---@return function # iterator function
+	function table.pairsByKeys(t, f)
+		local a = {}
+		for n in pairs(t) do table.insert(a, n) end
+		table.sort(a, f)
+		local i = 0           -- iterator variable
+		local iter = function() -- iterator function
+			i = i + 1
+			if a[i] == nil then
+				return nil
+			else
+				return a[i], t[a[i]]
+			end
+		end
+		return iter
+	end
```

Comment:
It is possible to give these variables (other than iterators) more meaningful names than single letter identifiers?

Reply:
This is the single I just straight copied from the official docs, but I can probably try to rename them yes.

Reply:
Done.